sources/ghostdog/src/main.rs

```diff
@@ -319,6 +379,302 @@ lazy_static! {
     ].iter().copied().collect();
 }
 
+#[derive(Deserialize, Debug)]
+struct InfinibandPort {
+    state: String,
+    #[serde(rename = "Physical state")]
+    physical_state: String,
+    rate: u32,
+    #[serde(rename = "Base lid")]
+    base_lid: u32,
+    lmc: u32,
+    #[serde(rename = "SM lid")]
+    sm_lid: u32,
+    #[serde(rename = "Capability mask")]
+    capability_mask: String, // Using String for hex value
+    #[serde(rename = "Port GUID")]
+    port_guid: String,
+    #[serde(rename = "Link layer")]
+    link_layer: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct InfinibandDevice {
+    #[allow(dead_code)]
+    name: String,
+    #[serde(rename = "CA type")]
+    ca_type: String,
+    #[serde(rename = "Number of ports")]
+    num_ports: u32,
+    #[serde(rename = "Firmware version")]
+    firmware_version: String,
+    #[serde(rename = "Hardware version")]
+    hardware_version: u32,
+    #[serde(rename = "Node GUID")]
+    node_guid: String,
+    #[serde(rename = "System image GUID")]
+    system_image_guid: String,
+    ports: Vec<InfinibandPort>,
+}
+
+fn hex_string_to_u32(hex: &str) -> std::result::Result<u32, std::num::ParseIntError> {
```

Comment:
helpful nit: You can redefine the `Result` type in this module as:

```rust
type Result<T, E = error::Error> = std::result::Result<T, E>;
```

Which would then let you express this as:

```suggestion
fn hex_string_to_u32(hex: &str) -> Result<u32, std::num::ParseIntError>
```